spot-contracts/contracts/vaults/RolloverVault.sol

```diff
@@ -315,6 +338,99 @@ contract RolloverVault is
+    /// @return The amount of underlying tokens returned.
+    function meld(IBondController bond, uint256[] memory trancheAmtsIn)
+        external
+        nonReentrant
+        whenNotPaused
+        returns (uint256)
+    {
+        // get bond tranches
+        BondTranches memory bt = bond.getTranches();
+
+        // 1) We check that the bond's underlying collateral token matches the vault's underlying.
+        // 2) We validate that the bond has at least 2 tranches to perform the meld operation.
+        // 3) We validate that the bond hasn't reached maturity as matching only works through immature redemption.
+        // 4) We check if the given bond is NOT malicious, by examining if at least one of the bond's tranches
+        //    is already in the vault from previous deployments. And if the parent bond of the tranche already in the vault
+        //    is the given bond.
+        if (
+            bond.collateralToken() != address(underlying) ||
+            bt.tranches.length <= 1 ||
+            bond.isMature() ||
+            !_isDeployedBond(bond, bt)
+        ) {
+            revert InvalidBond(bond);
+        }
+
+        // track the vault's TVL before the match operation
+        uint256 tvlBefore = getTVL();
+
+        // calculate total amount to be used by summing available amount from the user and the vault's balance
+        uint256[] memory trancheAmtsUsed = new uint256[](bt.tranches.length);
+        for (uint8 i = 0; i < bt.tranches.length; i++) {
+            trancheAmtsUsed[i] = trancheAmtsIn[i] + bt.tranches[i].balanceOf(address(this));
+        }
+
+        // compute proportional tranche amounts which can be redeemed for the underlying collateral
+        trancheAmtsUsed = bt.computeRedeemableTrancheAmounts(trancheAmtsUsed);
+
+        // computes the collateralizations for each tranche in the current immature bond
+        uint256[] memory trancheUnderlyingBalances;
+        uint256[] memory trancheSupplies;
+        (trancheUnderlyingBalances, trancheSupplies) = bond.getImmatureTrancheCollateralizations(bt);
+
+        // the user's share of the underlying collateral to be returned
+        uint256 underlyingAmt = 0;
+        for (uint8 i = 0; i < bt.tranches.length; i++) {
+            // compute the user's tranche balance to be user
+            // transfer tranche balance out of the user's wallet to the vault
+            trancheAmtsIn[i] = MathUpgradeable.min(trancheAmtsIn[i], trancheAmtsUsed[i]);
+            IERC20Upgradeable(bt.tranches[i]).safeTransferFrom(_msgSender(), address(this), trancheAmtsIn[i]);
+
+            // pre-calculate the user's share
+            underlyingAmt += trancheUnderlyingBalances[i].mulDiv(trancheAmtsIn[i], trancheSupplies[i]);
+        }
+
+        // ensure that the user is sending in tranches with some redeemable value.
+        if (underlyingAmt <= 0) {
```

Comment:
This is an interesting case...

Say the user is sending in Z's that are wiped out. It technically /should/ work... It's just that the user would get nothing back, and the Vault would be holding onto the underlying from its As share.

It would be weird for someone to do this, unless they were being charitable, and especially since we expect people to be sending in As. But sending in a Z worth 0 is not that much different from a Z that's worth some positive dust amount.

Reply:
This also prevents against the user sending in no tranches as all.

Reply:
True. Also, the vault wouldn't receive any fee for it. The risk might not be worth it for the small amount of short term liquidity that it might possibly (though very rarely, if ever) receive.

Reply:
Alternatively, the fee could be charged as a % of the incoming asset